scheduler/reconcile_util.go

```diff
@@ -215,12 +215,15 @@ func (a allocSet) fromKeys(keys ...[]string) allocSet {
 // 3. Those that exist on lost nodes
 // 4. Those that are on nodes that are disconnected, but have not had their ClientState set to unknown
 // 5. Those that have had their ClientState set to unknown, but their node has reconnected.
-func (a allocSet) filterByTainted(taintedNodes map[string]*structs.Node) (untainted, migrate, lost, disconnecting, reconnecting allocSet) {
+func (a allocSet) filterByTainted(taintedNodes map[string]*structs.Node, supportsDisconnectedClients bool) (untainted, migrate, lost, disconnecting, reconnecting allocSet) {
 	untainted = make(map[string]*structs.Allocation)
 	migrate = make(map[string]*structs.Allocation)
 	lost = make(map[string]*structs.Allocation)
-	disconnecting = make(map[string]*structs.Allocation)
-	reconnecting = make(map[string]*structs.Allocation)
+
+	if supportsDisconnectedClients {
+		disconnecting = make(map[string]*structs.Allocation)
+		reconnecting = make(map[string]*structs.Allocation)
+	}
```

Comment:
After a cluster is fully upgraded we'll take this branch 100% of the time. It feels like the "optimization" of not allocating these empty maps for this narrow case is going to be outweighed by any possibility we screw it up and then end up with a panic on assignment to a nil map.

Reply:
Fixed in latest push.